zerver/tests/test_auth_backends.py

```diff
@@ -1499,6 +1501,143 @@ def test_github_oauth2_success_single_email(self) -> None:
+        uri = "{}://{}{}".format(parsed_url.scheme, parsed_url.netloc,
+                                 parsed_url.path)
+        self.assertTrue(uri.startswith('http://zulip.testserver/accounts/login/subdomain/'))
+
+    def test_github_oauth2_login_no_account_exists(self) -> None:
+        # In login flow if the user has only one account and there are
+        # multiple verified emails, the choose email screen should be
+        # shown.
+        account_data_dict = dict(email='not-hamlet@zulip.com', name="Not Hamlet")
+        email_data = [
+            dict(email=account_data_dict["email"],
+                 verified=True),
+            dict(email="notprimary@zulip.com",
+                 verified=True),
+            dict(email="verifiedemail@zulip.com",
+                 verified=True),
+        ]
+        result = self.social_auth_test(account_data_dict,
+                                       subdomain='zulip',
+                                       email_data=email_data,
+                                       expect_choose_email_screen=True,
+                                       next='/user_uploads/image')
+        data = load_subdomain_token(result)
+        self.assertEqual(data['email'], account_data_dict["email"])
+        self.assertEqual(data['name'], account_data_dict["name"])
+        self.assertEqual(data['subdomain'], 'zulip')
+        self.assertEqual(data['next'], '/user_uploads/image')
+        self.assertEqual(result.status_code, 302)
+        parsed_url = urllib.parse.urlparse(result.url)
+        uri = "{}://{}{}".format(parsed_url.scheme, parsed_url.netloc,
+                                 parsed_url.path)
+        self.assertTrue(uri.startswith('http://zulip.testserver/accounts/login/subdomain/'))
+
+    def test_github_oauth2_signup_choose_existing_account(self) -> None:
+        # In sign up flow if the user has chooses an email of an
+        # existing account, the user gets an error saying account
+        # already exists for chosen email
+        account_data_dict = self.get_account_data_dict(email=self.email, name=self.name)
+        email_data = [
+            dict(email=account_data_dict["email"],
+                 verified=True),
+            dict(email="notprimary@zulip.com",
+                 verified=True),
+            dict(email="verifiedemail@zulip.com",
+                 verified=True),
+        ]
+        result = self.social_auth_test(account_data_dict,
+                                       email_data=email_data,
+                                       is_signup='1',
+                                       expect_choose_email_screen=True,)
+        self.assertEqual(result.status_code, 302)
+        result = self.client_get(result.url)
+
+        # Verify that we're unable to signup, since user chooses email of
+        # existing acount.
+        self.assertEqual(result.status_code, 200)
+        self.assert_in_success_response(["Sign up"], result)
```

Comment:
So in 27b9eafcac8825d84444c92c4713b1aa0aab59e9 we changed the behavior in the flow a bit, which means we'll now simply log the user in here, rather than going to the error. Perhaps that's actually better behavior. @timabbott  Do you agree with this? Or should we go to an error when a user picks an already registered email in the "Choose email" screen in the signup flow?

If we stick with the current state of master, this test should be adjusted to reflect that the user will be logged in here.

Reply:
I also think that's a better behaviour. will change the test.